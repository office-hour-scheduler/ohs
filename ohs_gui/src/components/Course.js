import React, { Component } from 'react';
import LectureSectionCard from "./dashboard/LectureSectionCard"
import { Link } from 'react-router-dom'

class Course extends Component {
  constructor(props) {
    super(props);
    this.state = {
      course: null,
      sections: []
    }

    this.getSections = this.getSections.bind(this);
    this.getCourse = this.getCourse.bind(this);
  }

  componentDidMount() {
    this.getCourse();
    this.getSections();
  }

  getSections() {
    // TODO: dummy json
    const sections = [
      {
        year: 2019,
        semester: "Winter",
        numStudents: 200,
        id: 6
      }, {
        year: 2019,
        semester: "Summer",
        numStudents: 200,
        id: 7
      }, {
        year: 2019,
        semester: "Fall",
        numStudents: 200,
        id: 8
      }
    ]
    this.setState({ sections: sections });
  }

  getCourse() {
    // TODO: dummy json
    const course = {
      name: "CSC302",
      id: this.props.match.params.id
    }
    this.setState({ course: course });
  }

  render() {
    const { course, sections } = this.state;
    return (
      <div id="sections">
        {course && <h2>{course.name} Lecture Sections</h2>}
        {sections.map(s => (
          <LectureSectionCard verbose section={s} key={s.id} />
        ))}
        {course && <Link to={`/course/${course.id}/addSection`}>
          <div className="add-section card-element">
            <span className="fa fa-plus"></span>
          </div>
        </Link>}
      </div>
    );
  }
}

export default Course;
