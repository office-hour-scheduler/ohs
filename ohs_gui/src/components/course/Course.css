#sections .container {
  width: 100%;
}

.header-cont {
  display: flex;
  align-items: center;
  justify-content: center;
}

.header-cont i {
  cursor: pointer;
}

i {
  transition: transform 0.1s, color 0.01s;
}

i:hover {
  transform: rotate(180deg);
}

i:active {
  color: red;
}
