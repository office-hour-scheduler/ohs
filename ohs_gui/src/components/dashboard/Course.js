import React, { Component } from 'react';

class Course extends Component {
  render() {
    return (
      <div className="course card-element">
        {this.props.course.name}
      </div>
    );
  }
}

export default Course;
