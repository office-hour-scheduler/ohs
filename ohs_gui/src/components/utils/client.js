import { ApolloClient } from 'apollo-client';
import { createHttpLink } from 'apollo-link-http';
import { setContext } from 'apollo-link-context';
import { InMemoryCache } from 'apollo-cache-inmemory';

const BASE_URL = `${process.env.REACT_APP_INSTRUCTOR_SERVICE_URL ||
  'http://localhost:8000'}`;

const getClient = () => {
  const httpLink = createHttpLink({
    uri: `${BASE_URL}/graphql`
  });

  const authLink = setContext((_, { headers }) => {
    // get the authentication token from local storage if it exists
    const token = window.sessionStorage.getItem('token');
    // return the headers to the context so httpLink can read them
    return {
      headers: {
        ...headers,
        Authorization: token ? `Bearer ${token}` : ''
      }
    };
  });

  const client = new ApolloClient({
    link: authLink.concat(httpLink),
    cache: new InMemoryCache()
  });

  return client;
};

const client = null;

export { client, BASE_URL, getClient };
