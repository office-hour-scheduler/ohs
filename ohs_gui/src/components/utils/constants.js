// User roles
const PROFESSOR = 1;
const STUDENT = 2;

export { PROFESSOR, STUDENT };
