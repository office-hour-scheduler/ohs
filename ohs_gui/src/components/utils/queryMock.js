// @flow
import { QueryMock } from 'graphql-query-test-mock';

const queryMock = new QueryMock();

export default queryMock;
