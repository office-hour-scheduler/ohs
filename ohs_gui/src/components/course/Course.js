/* eslint-disable react/forbid-prop-types */
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import shortid from 'shortid';
import { toast } from 'react-toastify';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';

import LectureSectionCard from '../dashboard/LectureSectionCard';
import { GET_SECTIONS_FOR_COURSE } from '../utils/queries';
import { userIsProf } from '../utils/helpers';

import './Course.css';

class Course extends Component {
  render() {
    const {
      user,
      match: {
        params: { courseCode }
      }
    } = this.props;

    const isProf = userIsProf(user);
    const variables = {
      courseCode,
      taughtBy: user.userName
    };

    return (
      <div id="sections">
        <div className="container">
          <div className="text-center">
            {courseCode && <h1>{courseCode} Lecture Sections</h1>}
          </div>
        </div>

        <Query
          query={GET_SECTIONS_FOR_COURSE}
          variables={variables}
          onError={() => {
            toast('Unknown Error - Could not get sections for the course', {
              type: toast.TYPE.ERROR
            });
          }}
        >
          {({ data }) => {
            if (data) {
              const { sections } = data;
              if (sections) {
                const lst = sections.map(s => (
                  <LectureSectionCard
                    verbose
                    section={s}
                    key={shortid.generate()}
                  />
                ));
                return lst;
              }
            }

            return null;
          }}
        </Query>

        {courseCode && isProf && (
          <Link to={`/course/${courseCode}/add-section`}>
            <div className="add-section card-element">
              <span className="fa fa-plus" />
            </div>
          </Link>
        )}
      </div>
    );
  }
}

Course.propTypes = {
  user: PropTypes.shape({
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    userName: PropTypes.string,
    role: PropTypes.number
  }).isRequired,
  // eslint-disable-next-line react/require-default-props
  match: PropTypes.object.isRequired
};

export default Course;
