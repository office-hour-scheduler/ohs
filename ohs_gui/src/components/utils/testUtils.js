import { roles } from './constants';

const PROF_USER = {
  firstName: 'Albert',
  lastName: 'Einstein',
  role: roles.PROFESSOR,
  userName: 'emc2'
};

const STUDENT_USER = {
  firstName: 'Bernie',
  lastName: 'Sanders',
  role: roles.STUDENT,
  userName: 'sensan'
};

const SECTIONS = [
  {
    course: {
      courseCode: 'CSC302'
    },
    year: 2019,
    semester: 'FALL',
    sectionCode: 'LEC0101',
    numStudents: 25
  },
  {
    course: {
      courseCode: 'CSC304'
    },
    year: 2019,
    semester: 'FALL',
    sectionCode: 'LEC0102',
    numStudents: 25
  }
];

export { SECTIONS, PROF_USER, STUDENT_USER };
