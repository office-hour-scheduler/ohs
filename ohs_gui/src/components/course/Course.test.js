import React from 'react';
import { shallow } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, cleanup } from 'react-testing-library';
import { MockedProvider } from 'react-apollo/test-utils';

import { STUDENT_USER, PROF_USER } from '../utils/testUtils';

import Course from './Course';

import queryMock from '../utils/queryMock';

beforeEach(() => {
  // Make sure no mocks stick around between tests
  queryMock.reset();
});

// Clean up any mounted DOM by react-testing-library
afterEach(cleanup);

const MATCH = { params: { courseCode: 'CSC302' } };

it('should warn correctly with user prop only', () => {
  const t = () =>
    shallow(
      <MockedProvider>
        <Course user={STUDENT_USER} />
      </MockedProvider>
    );
  expect(t).toThrow(Error);
});

it('should warn correctly with match prop only', () => {
  const t = () =>
    shallow(
      <MockedProvider>
        <Course match={MATCH} />
      </MockedProvider>
    );
  expect(t).toThrow(Error);
});

it('renders course name', () => {
  const { getByText } = render(
    <MockedProvider>
      <Router>
        <Course user={PROF_USER} match={MATCH} />
      </Router>
    </MockedProvider>
  );
  expect(
    getByText(`${MATCH.params.courseCode} Lecture Sections`)
  ).toBeDefined();
});

it('check if it prof renders properly', () => {
  const component = shallow(
    <MockedProvider>
      <Course user={PROF_USER} match={MATCH} />
    </MockedProvider>
  );
  expect(component).toMatchSnapshot();
});

it('check if it student renders properly', () => {
  const component = shallow(
    <MockedProvider>
      <Course user={STUDENT_USER} match={MATCH} />
    </MockedProvider>
  );
  expect(component).toMatchSnapshot();
});

// it('mock getting new sections', async () => {
//   // intercept http requests
//   queryMock.mockQuery({
//     name: 'getSectionForCourse',
//     data: {
//       sections: SECTIONS
//     },
//     variables: {
//       courseCode: MATCH.params.courseCode,
//       taughtBy: PROF_USER.userName
//     }
//   });

//   const wrapper = shallow(<MockedProvider><Course user={PROF_USER} match={MATCH} /></MockedProvider>);
//   wrapper.update();
//   expect(wrapper.state('sections').length).toEqual(2);
// });
