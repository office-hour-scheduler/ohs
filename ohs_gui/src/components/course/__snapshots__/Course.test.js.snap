// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check if it prof renders properly 1`] = `
<ApolloProvider
  client={
    ApolloClient {
      "cache": InMemoryCache {
        "addTypename": true,
        "cacheKeyRoot": CacheKeyNode {
          "children": null,
          "key": null,
        },
        "config": Object {
          "addTypename": true,
          "dataIdFromObject": [Function],
          "fragmentMatcher": HeuristicFragmentMatcher {},
          "resultCaching": true,
        },
        "data": DepTrackingCache {
          "data": Object {},
          "depend": [Function],
        },
        "maybeBroadcastWatch": [Function],
        "optimisticData": DepTrackingCache {
          "data": Object {},
          "depend": [Function],
        },
        "silenceBroadcast": false,
        "storeReader": StoreReader {
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "executeSelectionSet": [Function],
          "executeStoreQuery": [Function],
        },
        "storeWriter": StoreWriter {},
        "typenameDocumentCache": Map {},
        "watches": Set {},
      },
      "clearStoreCallbacks": Array [],
      "clientAwareness": Object {},
      "defaultOptions": Object {},
      "disableNetworkFetches": false,
      "link": ApolloLink {
        "request": [Function],
      },
      "localState": LocalState {
        "cache": InMemoryCache {
          "addTypename": true,
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "resultCaching": true,
          },
          "data": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "maybeBroadcastWatch": [Function],
          "optimisticData": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "silenceBroadcast": false,
          "storeReader": StoreReader {
            "cacheKeyRoot": CacheKeyNode {
              "children": null,
              "key": null,
            },
            "executeSelectionSet": [Function],
            "executeStoreQuery": [Function],
          },
          "storeWriter": StoreWriter {},
          "typenameDocumentCache": Map {},
          "watches": Set {},
        },
        "client": [Circular],
      },
      "mutate": [Function],
      "query": [Function],
      "queryDeduplication": true,
      "reFetchObservableQueries": [Function],
      "resetStore": [Function],
      "resetStoreCallbacks": Array [],
      "ssrMode": false,
      "store": DataStore {
        "cache": InMemoryCache {
          "addTypename": true,
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "resultCaching": true,
          },
          "data": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "maybeBroadcastWatch": [Function],
          "optimisticData": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "silenceBroadcast": false,
          "storeReader": StoreReader {
            "cacheKeyRoot": CacheKeyNode {
              "children": null,
              "key": null,
            },
            "executeSelectionSet": [Function],
            "executeStoreQuery": [Function],
          },
          "storeWriter": StoreWriter {},
          "typenameDocumentCache": Map {},
          "watches": Set {},
        },
      },
      "typeDefs": undefined,
      "version": "2.5.1",
      "watchQuery": [Function],
    }
  }
>
  <Course
    match={
      Object {
        "params": Object {
          "courseCode": "CSC302",
        },
      }
    }
    user={
      Object {
        "firstName": "Albert",
        "lastName": "Einstein",
        "role": 1,
        "userName": "emc2",
      }
    }
  />
</ApolloProvider>
`;

exports[`check if it student renders properly 1`] = `
<ApolloProvider
  client={
    ApolloClient {
      "cache": InMemoryCache {
        "addTypename": true,
        "cacheKeyRoot": CacheKeyNode {
          "children": null,
          "key": null,
        },
        "config": Object {
          "addTypename": true,
          "dataIdFromObject": [Function],
          "fragmentMatcher": HeuristicFragmentMatcher {},
          "resultCaching": true,
        },
        "data": DepTrackingCache {
          "data": Object {},
          "depend": [Function],
        },
        "maybeBroadcastWatch": [Function],
        "optimisticData": DepTrackingCache {
          "data": Object {},
          "depend": [Function],
        },
        "silenceBroadcast": false,
        "storeReader": StoreReader {
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "executeSelectionSet": [Function],
          "executeStoreQuery": [Function],
        },
        "storeWriter": StoreWriter {},
        "typenameDocumentCache": Map {},
        "watches": Set {},
      },
      "clearStoreCallbacks": Array [],
      "clientAwareness": Object {},
      "defaultOptions": Object {},
      "disableNetworkFetches": false,
      "link": ApolloLink {
        "request": [Function],
      },
      "localState": LocalState {
        "cache": InMemoryCache {
          "addTypename": true,
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "resultCaching": true,
          },
          "data": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "maybeBroadcastWatch": [Function],
          "optimisticData": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "silenceBroadcast": false,
          "storeReader": StoreReader {
            "cacheKeyRoot": CacheKeyNode {
              "children": null,
              "key": null,
            },
            "executeSelectionSet": [Function],
            "executeStoreQuery": [Function],
          },
          "storeWriter": StoreWriter {},
          "typenameDocumentCache": Map {},
          "watches": Set {},
        },
        "client": [Circular],
      },
      "mutate": [Function],
      "query": [Function],
      "queryDeduplication": true,
      "reFetchObservableQueries": [Function],
      "resetStore": [Function],
      "resetStoreCallbacks": Array [],
      "ssrMode": false,
      "store": DataStore {
        "cache": InMemoryCache {
          "addTypename": true,
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "resultCaching": true,
          },
          "data": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "maybeBroadcastWatch": [Function],
          "optimisticData": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "silenceBroadcast": false,
          "storeReader": StoreReader {
            "cacheKeyRoot": CacheKeyNode {
              "children": null,
              "key": null,
            },
            "executeSelectionSet": [Function],
            "executeStoreQuery": [Function],
          },
          "storeWriter": StoreWriter {},
          "typenameDocumentCache": Map {},
          "watches": Set {},
        },
      },
      "typeDefs": undefined,
      "version": "2.5.1",
      "watchQuery": [Function],
    }
  }
>
  <Course
    match={
      Object {
        "params": Object {
          "courseCode": "CSC302",
        },
      }
    }
    user={
      Object {
        "firstName": "Bernie",
        "lastName": "Sanders",
        "role": 2,
        "userName": "sensan",
      }
    }
  />
</ApolloProvider>
`;
